### PR TITLE
Update SpaCy to 2.0

### PR DESCRIPTION
**Pull request by nathanathan**
This required the following changes:

1. The lemmatized forms of some words searched for in the case count annotator were altered.
2. The token similarity function can be used for word sense disambiguation, so sentence similarity is no longer used.
3. Some tests were sensitive to false positives in the NER. Minor tweaks were made to their texts to avoid false positives.
4. The geoname classifier needed to be retrained since it relies on SpaCy features. To maintain classifier performance I removed the min_token_prob feature, added the combined_span_parents feature which is the number of unique parent geonames given in combined spans like "Seattle, WA", altered the cannonical_name_used feature so that partial matches and asciiname matches are scored, and fixed a bug affecting span based features where combined spans should have been replacing their source span.

This also adds a structured data annotator which identifies tables and key-value lists embedded in documents and extracts them as python data structures.

### CODE REVIEW

File: epitator/geoname_classifier.py [hunk 1]
```diff
@@ -12,46 +12,48 @@
...
  'classes_': array([False,  True], dtype=bool),
- 'coef_': array([[ 0.30886707,  0.02116973,  0.00560847,  0.27765891,  1.94456142,
-         0.52508845,  0.18140931,  0.        , -0.95166464,  0.01895766,
-        -0.01038021, -0.7187989 , -0.24003574,  2.73676538,  0.        ,
-        -0.12539797,  0.        ,  0.        ,  0.        ,  0.        ,
+ 'coef_': array([[ 0.3213819 ,  0.01823242,  0.02336938,  0.29849362,  2.47665562,
+         0.47210803,  0.16109926,  0.        , -0.97918471,  0.00476336,
+        -0.01442839, -0.60016267, -0.23542543,  1.89213763,  0.        ,
+        -1.68017007,  0.        ,  0.        ,  0.        ,  0.        ,
          0.        ]]),
```
Comment (arnoldbm) [inline epitator/geoname_classifier.py]: How did we determine these new values?

File: epitator/count_annotator.py [hunk 2]
```diff
@@ -196,29 +184,16 @@ def search_lemmas(lemmas, match_name=None):
             max_dist=50)
         # Add singular case reports
         singular_case_spans = []
-        determiner_lemmas = set(['a', 'an', 'the'])
+        determiner_lemmas = set(['a', 'an', 'the', 'one'])
```
Comment (arnoldbm) [inline epitator/count_annotator.py]: Would be interesting to test 'each' here also.

File: epitator/geoname_annotator.py [hunk 3]
```diff
@@ -211,9 +213,17 @@ def __init__(self, geoname, spans_to_nes, span_to_tokens):
...
-        d['cannonical_name_used'] = 1 if any([
-            span.text == geoname.name for span in geoname.spans
-        ]) else 0
+        def cannonical_name_match(span, geoname):
+            if hasattr(span, "iterate_leaf_base_spans"):
+                span_text = next(span.iterate_leaf_base_spans()).text
+            else:
+                span_text = span.text
+            span_in_name = span_text in geoname.name or span_text in geoname.asciiname
+            return (float(len(span_text)) if span_in_name else 0) / len(geoname.name)
```
Comment (arnoldbm) [inline epitator/geoname_annotator.py]: Should we have a hasattr on the name for geoname here, to make sure we don't divide by zero?

File: epitator/geoname_classifier.py (hunk 1 above)
Comment (nathanathan) [reply epitator/geoname_classifier.py]: The training script at https://github.com/ecohealthalliance/geoname-annotator-training

File: epitator/geoname_annotator.py (hunk 3 above)
Comment (nathanathan) [reply epitator/geoname_annotator.py]: name is a required attribute for all geonames.

File: epitator/count_annotator.py (hunk 2 above)
Comment (nathanathan) [reply epitator/count_annotator.py]: What context were you thinking of "each" appearing in? I'm not seeing how that would apply to singular case reports.

File: epitator/geoname_classifier.py (hunk 1 above)
Comment (toph-allen) [reply epitator/geoname_classifier.py]: Do we know what effect this had on annotator performance, or are we not doing cross-validation on the GeoName annotator?

Comment (nathanathan) [reply epitator/geoname_classifier.py]: There's no cross validation, but the geoname annotator is scored against the training data. There appears to be a drop in accuracy. I think it's worth taking since we'll be on stronger footing moving forward with SpaCy 2. Here's the current and previous stats:

```
current:
tps: 1223
fns: 451
fps: 487
ignored: 247

previous:
tps: 1272
fns: 402
fps: 425
ignored: 246
```

File: epitator/structured_data_annotator.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            }))
+        for token, start, end in key_value_list_parser.scanString(doc.text):
+            spans.append(AnnoSpan(start, end, doc, "keyValuePairs", metadata={
+                "type": "keyValuePairs",
+                "data": {
+                    pair[0]: pair[1]
+                    for pair in token
+                }
+            }))
+        return {'structuredData': AnnoTier(spans)}
```
Comment (toph-allen) [inline epitator/structured_data_annotator.py]: This seems like a good first pass. One thing to note: There's another example of a key-value pair in the [Slack discussion of exported annotations]()https://eha.slack.com/archives/G0P7PPUAW/p1511302007000117):
>The number of ill people identified in each state with the outbreak strain is as follows: Alabama (2), Arkansas (1), Connecticut (4), District of Columbia (2), Georgia (4), Illinois (8), Louisiana (2), Maryland (8), Massachusetts (4), Mississippi (1), Missouri (1), New Jersey (6), New York (23), North Carolina (2), Pennsylvania (2), Rhode Island (4), South Carolina (3), Texas (3), Virginia (5), and Wisconsin (8).

I wonder if it'd be useful broadly to also flag non-natural-language regions so that our NLP algorithms don't go haywire over them. I'm thinking of things in addition to tables and key-value lists — things like front matter etc.

But for now — I think this is a good first pass, and I'm interested to see how it operates. We should probably not necessarily use it to create Incidents yet, b/c I'm sure that the way we usually assign dates and locations and other metadata to them won't necessarily apply to this stuff.

File: tests/annotator/test_structured_data_annotator.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                'TypeError County': '1 case',
+                'Python County': '2 cases',
+                'Java County': '2 cases',
+                'Scala County': '1 case',
+                'Scheme County': '1 case',
+                'Meteor County': '1 case',
+                'Boolean County': '1 case (not including the fatality)',
+                'Integer County': '3 cases'
+            }
+        })
```
Comment (toph-allen) [inline tests/annotator/test_structured_data_annotator.py]: 👍👍👍

**Comment by toph-allen** [comment]
Thanks! Sounds good. We’ll run CV at some point this year… :D

--

Toph Allen, MPH
Director of Data Science

EcoHealth Alliance
460 West 34th Street, 17th Floor
New York NY 10001

+1.213.448.4152 (mobile)
+1.212.380.4481 (direct)
+1.212.380.4465 (fax)
www.ecohealthalliance.org

EcoHealth Alliance leads cutting-edge research into the critical connections between human and wildlife health and delicate ecosystems. With this science, we develop solutions that prevent pandemics and promote conservation.
